File: osx/ColorInspector/ColorInspector/MouseHook.m
//
//  MouseHook.m
//  ColorInspector
//
//  Created by Alan Rahlf on 8/9/15.
//  Copyright (c) 2015 Alan Rahlf. All rights reserved.
//

#import "MouseHook.h"

@interface MouseHook()

@property (assign, nonatomic) CFMachPortRef eventTap;

@end

@implementation MouseHook

- (id)init {
    if (self = [super init]) {
        [self setupMouseHook];
    }
    return self;
}

CGEventRef handleCGEvent(CGEventTapProxy proxy, CGEventType type, CGEventRef eventRef, void *refcon)
{
    MouseHook *mouseHook = (__bridge MouseHook*)refcon;
    
    if (type == kCGEventMouseMoved) {
        [mouseHook.delegate onMouseMove:CGEventGetLocation(eventRef)];
    }
    else if (type == kCGEventLeftMouseUp) {
        [mouseHook.delegate onMouseUp:CGEventGetLocation(eventRef)];
    }
    else if (type == kCGEventTapDisabledByTimeout) {
        CGEventTapEnable(mouseHook.eventTap, true);
    }
    
    return eventRef;
}

- (void)setupMouseHook {
    self.eventTap = CGEventTapCreate(kCGHIDEventTap, kCGHeadInsertEventTap, kCGEventTapOptionListenOnly, kCGEventMaskForAllEvents, handleCGEvent, (__bridge void *)(self));
    CFRunLoopSourceRef runLoopSource = CFMachPortCreateRunLoopSource(kCFAllocatorDefault, self.eventTap, 0);
    CFRunLoopAddSource(CFRunLoopGetMain(), runLoopSource, kCFRunLoopCommonModes);
}

@end

File: osx/ColorInspector/ColorInspector/MouseHook.swift
//
//  MouseHook.swift
//  ColorInspector
//
//  Created by Alan Rahlf on 11/1/15.
//  Copyright © 2015 Alan Rahlf. All rights reserved.
//

import Cocoa

@objc protocol MouseHookDelegate {
    func onMouseMove(point: CGPoint);
    func onMouseUp(point: CGPoint);
}

func handleCGEvent(proxy: CGEventTapProxy, type: CGEventType, event: CGEvent, refcon: UnsafeMutablePointer<Void>) -> Unmanaged<CGEvent>? {
    let mouseHook: MouseHook = Unmanaged.fromOpaque(COpaquePointer(refcon)).takeUnretainedValue();
    
    if (type == .MouseMoved || type == .LeftMouseDragged) {
        mouseHook.delegate?.onMouseMove(CGEventGetLocation(event));
    }
    else if (type == .LeftMouseUp) {
        mouseHook.delegate?.onMouseUp(CGEventGetLocation(event));
    }
    else if (type == .TapDisabledByTimeout) {
        CGEventTapEnable(mouseHook.eventTap!, true);
    }
    
    return Unmanaged.passRetained(event);
}

@objc class MouseHook: NSObject {
    
    var delegate: MouseHookDelegate?;
    
    private var eventTap: CFMachPortRef?;

    override init() {
        super.init();
        
        self.setupMouseHook();
    }
    
    func setupMouseHook() {
        let refcon = UnsafeMutablePointer<Void>(Unmanaged.passUnretained(self).toOpaque());
        
        self.eventTap = CGEventTapCreate(.CGHIDEventTap, .HeadInsertEventTap, .ListenOnly, ~CGEventMask(0), handleCGEvent, refcon);
        let runLoopSource = CFMachPortCreateRunLoopSource(kCFAllocatorDefault, self.eventTap, 0);
        CFRunLoopAddSource(CFRunLoopGetMain(), runLoopSource, kCFRunLoopCommonModes);
    }
}
